Add dev2 branch for RRFS data assimilation

hu5970 opened the pull request:
Added the branch to hold the data assimilation features. 
Three hourly cycling with cold start at 00Z and 12Z were added.


## DESCRIPTION OF CHANGES: 
1) New scripts files in jobs directory: JREGIONAL_PROCESS_RADARREF, JREGIONAL_REFL2TTEN, JREGIONAL_RUN_ANAL
2) New module and run  files in modulefiles directory
3) New scripts files in scripts directory: exregional_run_analysis.sh, exregional_refl2tten.sh,exregional_process_radarref.sh, and updated exregional_run_fcst.sh 
4) New configure file (config.sh.RRFS_dev2) and updated scripts in ush directory and templates directory.

## TESTS CONDUCTED: 
RRFS_dev2 is running in real time:
run location: /mnt/lfs4/BMC/nrtrr/NCO_dirs/stmp/tmpnwprd/RRFS_dev2
com location: /mnt/lfs4/BMC/nrtrr/NCO_dirs/ptmp/com

hu5970 committed:
Add data assimilation cycle to workflow, including: jobs, scripts, modules, xml generation

hu5970 committed:
Add data assimilation cycle information into var_defns.sh.

hu5970 committed:
Add scripts to generate the FV3LAM namelist for running forecast from restart files.

christinaholtNOAA commented on jobs/JREGIONAL_RUN_ANAL:
```diff
@@ -0,0 +1,103 @@
+#-----------------------------------------------------------------------
+#
+$SCRIPTSDIR/exregional_run_analysis.sh \
+  CYCLE_DIR="${CYCLE_DIR}" ANALWORKDIR="${ANALWORKDIR}" CYCLE_ROOT="${CYCLE_ROOT}" || \
```
Putting each argument on a separate line will make it easier to manage the code in the future.

hu5970 replied:
Done.

christinaholtNOAA commented on modulefiles/tasks/jet/radar_refl2tten:
```diff
@@ -0,0 +1 @@
+run_anal_gsi
```
I'd prefer that all of these be removed, and `run_anal_gsi` be called directly from the XML.

hu5970 replied:
Agree. Now, use run_anal_gsi for all analysis components from XML.

christinaholtNOAA commented on scripts/exregional_process_radarref.sh:
```diff
@@ -0,0 +1,326 @@
+#HH=17
+NSSL=${OBSPATH_NSSLMOSIAC}
+mrms="MRMS_EXP_MergedReflectivityQC"
+echo "${MM0} ${MM1} ${MM2} ${MM3}"
```
Where are these variables set?

hu5970 replied:
Those will be defined in "var_defns.sh".
This PR only add TTEN process as a place hold. Will add those steps in next PR.

christinaholtNOAA commented on scripts/exregional_process_radarref.sh:
```diff
@@ -0,0 +1,326 @@
+#
+#-----------------------------------------------------------------------
+
+#HH=17
```
Please remove unnecessary commented lines.

hu5970 replied:
done

christinaholtNOAA commented on scripts/exregional_process_radarref.sh:
```diff
@@ -0,0 +1,326 @@
+#
+export KMP_AFFINITY=scatter
+export OMP_NUM_THREADS=1 #Needs to be 1 for dynamic build of CCPP with GFDL fast physics, was 2 before.
+export OMP_STACKSIZE=1024m
```
Are these necessary for this application? The comment is misleading.

hu5970 replied:
removed

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+  ulimit -s unlimited
+  ulimit -a
+  APRUN="srun"
+#  LD_LIBRARY_PATH="${UFS_WTHR_MDL_DIR}/FV3/ccpp/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
```
Please remove unnecessary commented lines.

hu5970 replied:
Done

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+rectory for radar reflectivity process ..."
+
+
+# Create directory.
```
This comment is not true. The next two lines could be updated for clarity.
```
cd_vrfy ${WORKDIR}
```

hu5970 replied:
changed to "create a subdirectory..."

christinaholtNOAA replied:
This is the type of inconsistency that I'm seeing between the code `# Create directory` and you mentioned that you have changed it to read `# Create a subdirectory`.

hu5970 replied:
I see. I only changed the first line but the 2nd line need to be changed too.

christinaholtNOAA replied:
This still needs to be changed.

hu5970 replied:
Changed this time.

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+cp_vrfy ${fixdir}/fv3_grid_spec                          fv3_grid_spec
+
+bkpath=${CYCLE_DIR}/INPUT
+if [ -w ${bkpath}/gfs_data.tile7.halo0.nc ]; then  # Use background from INPUT
```
For clarity, can you leave a comment on what's happening with the "else" block?

hu5970 replied:
Added comment.

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+#  fi
+#  count=$((count + 1))
+#done
+ln -s ${PROCESS_RADARREF_PATH}/RefInGSI3D.dat ./RefInGSI3D.dat_01
```
Please remove commented code.

hu5970 replied:
Those are hold for the future if we do sub-hour or multiple time process.

christinaholtNOAA replied:
In general, it's best to add this type of feature when it is needed since there's a significant chance the logic could need to be re-worked when it's ready anyway. A comment that says something like "Need to link in subhourly RefInGSI3D.dat_0* files here" would be better.

hu5970 replied:
OK, I have removed this part.

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+# Build namelist and run GSI
+#
+#-----------------------------------------------------------------------
+
```
Please remove this comment section since this is not applicable for this application.

hu5970 replied:
Changed this comment.

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+# Set and export variables.
+#
+#-----------------------------------------------------------------------
+#
```
Please remove this comment section since this is not applicable for this application.

hu5970 replied:
Updated to be consistent with the application

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,297 @@
+# Run the GSI.  Note that we have to launch the forecast from
+# the current cycle's run directory because the GSI executable will look
+# for input files in the current directory.
+#
```
Please update comment to be consistent with application.

hu5970 replied:
updated

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+  ulimit -a
+  APRUN="srun"
+  LD_LIBRARY_PATH="${UFS_WTHR_MDL_DIR}/FV3/ccpp/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
+#  NCKS=/apps/nco/4.9.1/intel/18.0.5.274/bin/ncks
```
Please remove commented code.

hu5970 replied:
done

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+rectory to the grid and (filtered) orography files ..."
+
+
+# Create directory.
```
Please update comment.

hu5970 replied:
DONE

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+
+cd_vrfy ${ANALWORKDIR}
+
+#fixdir=$FIXgsi/$PREDEF_GRID_NAME
```
Please remove commented code.

hu5970 replied:
done

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+#${LS} ${ENKF_FCST}/${enkfcstname}.mem???.nemsio > filelist03
+
+# Check GDAS file size before creating file list
+#typeset -RZ3 n=002
```
Please remove commented code.

hu5970 replied:
done

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+beta1_inv=1.0
+ifhyb=.false.
+#nummem=80
+#fv3sar_bg_type=0
```
Please remove commented code.

hu5970 replied:
done

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+  bkpath=${CYCLE_DIR}/INPUT
+else
+  bkpath=${CYCLE_ROOT}/${YYYYMMDDHHmInterv}/RESTART
+fi
```
Please provide documentation on the new variable BKTYPE, along with its possible values and meanings.

hu5970 replied:
DONE

christinaholtNOAA replied:
It see a comment (well, an info message) in another script about the values. Could you elaborate in this script with a comment here about what each value of BKTYPE means in terms of being a cold start, cylced, etc.? Just grabbing info from INPUT or RESTART doesn't quite close that gap.

hu5970 replied:
Agree, I check all scripts and added "cold start" and "cycling" with the INPUT and RESTART to make it clear.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+Restart hour should not larger than forecast hour:
+    Restart Hour = \"${DA_CYCLE_INTERV}\"
+    Forecast Hour = \"${FCST_LEN_HRS}\""    
+    exit
```
Should this exit with a non-zero exit status?

hu5970 replied:
change to exit 1

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+#obsdir=${OBSPATH}/gfs.${YYYYMMDD}/${HH}
+#obs_file=${obsdir}/gfs.t${HH}z.prepbufr.nr
+#obs_file=${OBSPATH}/${YYYY}${JJJ}${HH}00.rap.t${HH}z.prepbufr.tm00.${YYYYMMDD}
+
```
Please remove commented code (here and in the next section)

hu5970 replied:
removed obs part but keep addtten part for now. Will remove it when figure out better way to make then array.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,618 @@
+export KMP_AFFINITY=scatter
+export OMP_NUM_THREADS=1 #Needs to be 1 for dynamic build of CCPP with GFDL fast physics, was 2 before.
+export OMP_STACKSIZE=1024m
+#
```
Are these necessary for this application? The comment is misleading.

hu5970 replied:
may be not. removed them.

christinaholtNOAA commented on scripts/exregional_run_fcst.sh:
```diff
@@ -26,7 +26,7 @@
 #-----------------------------------------------------------------------
 #
-{ save_shell_opts; set -u +x; } > /dev/null 2>&1
+{ save_shell_opts; set -u -x; } > /dev/null 2>&1
```
Is this only for debugging?

hu5970 replied:
Reverted.

christinaholtNOAA commented on ush/config.sh.RRFS_dev2:
```diff
@@ -0,0 +1,76 @@
+MACHINE="jet"
+ACCOUNT="nrtrr"
+EXPT_BASEDIR="/mnt/lfs4/BMC/rtwbl/mhu/rrfs/dev2"
```
This should be updated to the operational location.

hu5970 replied:
Done

christinaholtNOAA commented on ush/config.sh.RRFS_dev2:
```diff
@@ -0,0 +1,76 @@
+QUILTING="TRUE"
+USE_CCPP="TRUE"
+CCPP_PHYS_SUITE="FV3_GSD_SAR"
+FCST_LEN_HRS="12"
```
Will this be a longer forecast?

hu5970 replied:
The 12-h forecast now for cycling is enough. Will figure out how long the forecast need to be for each cycle in the future.

christinaholtNOAA commented on ush/config.sh.RRFS_dev2:
```diff
@@ -0,0 +1,76 @@
+
+HALO_BLEND=10
+
+WTIME_RUN_FCST="02:15:00"
```
This will need to be longer for a longer forecast.

hu5970 replied:
It is good for now. We only do long forecast in certain cycles in the future.

christinaholtNOAA commented on ush/config.sh.RRFS_dev2:
```diff
@@ -0,0 +1,76 @@
+COMINgfs=""
+
+STMP="/mnt/lfs4/BMC/rtwbl/mhu/rrfs/dev2/run/stmp"  # Path to directory STMP that mostly contains input files.
+PTMP="/mnt/lfs4/BMC/rtwbl/mhu/rrfs/dev2/run/ptmp"  # Path to directory STMP that mostly contains input files.
```
Please add the real-time locations.

hu5970 replied:
Done

christinaholtNOAA commented on ush/config_defaults.sh:
```diff
@@ -357,7 +357,10 @@ WFLOW_LAUNCH_LOG_FN="log.launch_FV3LAM_wflow"
 DATE_LAST_CYCL="YYYYMMDD"
 CYCL_HRS=( "HH1" "HH2" )
-FCST_LEN_HRS="24"
+FCST_LEN_HRS="12"
```
Please don't change this default.

hu5970 replied:
reverted

christinaholtNOAA commented on ush/config_defaults.sh:
```diff
@@ -357,7 +357,10 @@ WFLOW_LAUNCH_LOG_FN="log.launch_FV3LAM_wflow"
+FCST_LEN_HRS="12"
+RESTA_INTERV="3,6"  # check this
+DA_CYCLE_INTERV="3"
+restart_interval="3,6"
```
Please follow the all-caps convention for the global variables.

TrevorAlcott-NOAA replied:
What is the difference between "RESTA_INTERV" and "RESTART_INTERVAL"?  Might need a more descriptive variable name

hu5970 replied:
Added the descriptions and changed to all-caps convention.

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-
+set -x
```
Is this only for debugging?

hu5970 replied:
deleted

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.sh:
```diff
@@ -818,6 +823,33 @@ for the various ensemble members failed."
+Call to python script set_namelist.py to generate an FV3 namelist file
+failed.  Parameters passed to this script are:
+  Full path to base namelist file:
+    FV3_NML_BASE_SUITE_FP = \"${FV3_NML_BASE_SUITE_FP}\"
```
Please update this err msg to reflect the values used in the set_namelist command line args.

hu5970 replied:
Done

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -89,10 +98,13 @@ tasks; and the "FCST" type is used for the RUN_FCST_TN task.
 ]>
 
-<workflow realtime="T" scheduler="&SCHED;" cyclethrottle="3" cyclelifespan="01:00:00:00">
+<workflow realtime="T" scheduler="&SCHED;" cyclethrottle="30" cyclelifespan="01:00:00:00">
```
We do not want 30 cycles active in real time.

TrevorAlcott-NOAA replied:
Ideally cyclethrottle should be a function of the number of cycles per day and the longest deadline, not a hard-coded default. Can leave for now but we should tackle in a future PR

hu5970 replied:
Agree, this should be a parameter can be set based on the experiment. For the real time, should we have a big number so the job will run as needed. This number seems only make sense in retrospective experiments.

christinaholtNOAA replied:
The larger the number, the bigger the burden on the resources that run rocoto cron jobs.

hu5970 replied:
I reverted it back to 3.

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -89,10 +98,13 @@ tasks; and the "FCST" type is used for the RUN_FCST_TN task.
   <cycledef group="at_start">{{ cdate_first_cycl.strftime("%M %H %d %m %Y *") }}</cycledef>
   <cycledef group="forecast">{{ cdate_first_cycl.strftime("%Y%m%d%H%M") }} {{ cdate_last_cycl.strftime("%Y%m%d%H%M") }} {{ cycl_freq }}</cycledef>
+{%- if do_dacycle %}
+  <cycledef group="dacycle">00 03,06,09,15,18,21 {{cdate_first_cycl.strftime("%d") }}-{{ cdate_last_cycl.strftime("%d") }} {{ cdate_first_cycl.strftime("%m") }} 2020 *</cycledef>
```
Can we use the same syntax in this cycledef as in the forecast with 3 hour frequency?

hu5970 replied:
We need to define which cycle to cold start and which cycle to do continue cycle.
The syntax in forecast group only good for regular cycling. 
Should the forecast group use the syntax from dacycle? When setup real RRFS, we will not cold start from 00/12Z.

christinaholtNOAA replied:
I don't understand the details here. We could define cold starts on the "forecast" group and cycling on the "dacycle" group.

hu5970 replied:
The main issue here is continue cycling only run at 03, 06, 09, 15, 18, and 21Z. It is not regular 3 hour interval. Missing 00 and 12Z. So I have to define it explicitly.

christinaholtNOAA replied:
Oh! I definitely missed that! So we're running a cold start at 00 and 12Z?

hu5970 replied:
Yes, 00 and 12Z is cold start with analysis.

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -263,9 +279,9 @@ MODULES_RUN_TASK_FP script.
       <sh>
-        test $(find <cyclestr offset="-{{ extrn_mdl_lbcs_offset_hrs }}:00:00">{{ extrn_mdl_sysbasedir_lbcs }}/@Y@m@d@H/postprd</cyclestr> \
-               -name wrfnat_rr_[0-3][0-9].grib2 -mmin +5 | wc -l) \
-        == 40
```
We do need all the LBCs available for the full forecast time here.

hu5970 replied:
We only need 18 hours for the data assimilation. Those checks should be set based on the forecast length in the future. Or just check the last forecast needed for the boundary conditions based on the forecast length. Also. We may need to separate boundary generation and cycling in the future.

hu5970 replied:
This part should be controlled by the configuration.

christinaholtNOAA replied:
Can you please add a template section for that, then? This will surely get missed when we update the forecast length, otherwise.

hu5970 replied:
I think to check the longest forecast result file (single file) is good enough here.
I reverted to original value. Will change this check in later PR.
Also, we will need prepare much longer boundary condition than the forecast length when use GFS.

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -369,11 +389,84 @@ MODULES_RUN_TASK_FP script.
+
+    &RSRV_DEFAULT;
+    &WALL_LIMIT_ANAL;
+    <!--&WALL_LIMIT_WHOLE;-->
```
Please remove commented line

hu5970 replied:
Done

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -369,11 +389,84 @@ MODULES_RUN_TASK_FP script.
+    <command>&LOAD_MODULES_RUN_TASK_FP; "&ANAL_GSI_INPUT_TN;" "&JOBSDIR;/JREGIONAL_RUN_ANAL"</command>
+
+    <nodes>16:ppn=24</nodes>
+    <walltime>00:30:00</walltime>
```
Please use Jinja template here with values from the config settings.

hu5970 replied:
Done

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -369,11 +389,84 @@ MODULES_RUN_TASK_FP script.
+
+    &RSRV_DEFAULT;
+    &WALL_LIMIT_ANAL;
+    <!--&WALL_LIMIT_WHOLE;-->
```
Please remove commented line.

hu5970 replied:
removed

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -369,11 +389,84 @@ MODULES_RUN_TASK_FP script.
+    &WALL_LIMIT_ANAL;
+    <!--&WALL_LIMIT_WHOLE;-->
+
+    <command>&LOAD_MODULES_RUN_TASK_FP; "&ANAL_GSI_INPUT_TN;" "&JOBSDIR;/JREGIONAL_RUN_ANAL"</command>
```
Following on from my comment about the modulefiles, you can just put "anal_gsi_input" here in stead of `ANAL_GSI_INPUT_TN` to load the appropriate module.

hu5970 replied:
Yes, changed

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -402,8 +495,16 @@ MODULES_RUN_TASK_FP script.
   
     <dependency>
       <and>
+        <or>
```
<or> needs to be inside if statement.

hu5970 replied:
Yes, done.

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -489,6 +594,8 @@ MODULES_RUN_TASK_FP script.
 ************************************************************************
 -->
+
+<!--
```
These will need to be remove to deliver NCL graphics to the web. Please also address the commented tasks below.

hu5970 replied:
uncommented "ARCHIVE_TN" and "RUN_NCL_ZIP_TN".

christinaholtNOAA reviewed (changes requested):
Ming, I've left a bunch of comments here. Mostly about cleaning up the code for ease of management later on. In general, we don't want to hold onto commented code moving forward.

hu5970 committed:
Revise scripts based on Christina's comments.

hu5970 commented:
Christina, 
Thanks for comments. Very helpful.
Please take your time to review them again.

After your review and before it is ready for merge. I would like to test the cycling again to make sure everything works.

Thanks,
Ming

christinaholtNOAA commented:
@hu5970, thanks for the quick turn around on my comments!

Did you see all of my comments on the first review? There are some that are still unaddressed.

hu5970 committed:
Updated more scripts absed on Christina and Trevor's comments.

hu5970 commented:
Christina and Trevor,

Now I think I addressed all the comments. 
Please check and resolve those comments.

We can follow up with the unresolved ones after thanksgiving.

Thanks,
Ming

christinaholtNOAA reviewed (commented):
Ming, please push your latest changes. The code doesn't seem consistent with your comments.

hu5970 commented:
Christina,

I have push my code changes to my fork. But the PR still show the old code unless you pick the latest commit 205d11a.
Seems the GitHub doesn't automatically show the updated code.

Other than the three comments you listed, if you still find the code is not consistent with the my answer, please give me more details so I can double check.

Thanks,
Ming

hu5970 committed:
Update code based on Christina's review.

christinaholtNOAA reviewed (commented):
Ming, thanks for the additional updates. I'm also seeing that there are quite a few comments in scripts/exregional_run_analysis.sh that have not been addressed.

hu5970 committed:
Update scripts/exregional_run_analysis.sh based on Christina's comments.

hu5970 commented:
Christina,

Any more comments?

Thanks,
Ming

christinaholtNOAA commented:
I'd like to go over it with fresh eyes tomorrow.

hu5970 committed:
Clean exregional_process_radarref.sh

christinaholtNOAA commented on jobs/JREGIONAL_PROCESS_RADARREF:
```diff
@@ -0,0 +1,103 @@
+# directory, creates links to some of them, and modifies others (e.g. 
+# templates) to customize them for the current experiment setup.
+#
+#-----------------------------------------------------------------------
```
This boiler plate section here (and in the other jobs scripts) does not reflect the work being done by the script. Please consider making each of them more descriptive.

hu5970 replied:
I changed this part in three new J-jobs to just say what this J-JOB is for.

christinaholtNOAA commented on jobs/JREGIONAL_PROCESS_RADARREF:
```diff
@@ -0,0 +1,103 @@
+#
+# Create the INPUT and RESTART directories under the cycle directory.
+#
+#-----------------------------------------------------------------------
```
This comment is not accurate.

hu5970 replied:
Change to "Create the working directory under the cycle directory"

christinaholtNOAA commented on jobs/JREGIONAL_REFL2TTEN:
```diff
@@ -0,0 +1,103 @@
+#
+# Create the INPUT and RESTART directories under the cycle directory.
+#
+#-----------------------------------------------------------------------
```
This comment is not accurate.

hu5970 replied:
Change to "Create the working directory under the cycle directory"

christinaholtNOAA commented on jobs/JREGIONAL_RUN_ANAL:
```diff
@@ -0,0 +1,105 @@
+#
+# Create the INPUT and RESTART directories under the cycle directory.
+#
+#-----------------------------------------------------------------------
```
This comment is not accurate.

hu5970 replied:
Change to "Create the working directory under the cycle directory"

christinaholtNOAA commented on scripts/exregional_process_radarref.sh:
```diff
@@ -0,0 +1,312 @@
+rectory for radar reflectivity process ..."
+
+
+# Create directory.
```
This comment is not accurate.

hu5970 replied:
Changed to simply say "get into working directory".

christinaholtNOAA commented on scripts/exregional_process_radarref.sh:
```diff
@@ -0,0 +1,312 @@
+FV3SARPATH=${CYCLE_DIR}
+cp_vrfy ${fixdir}/fv3_grid_spec          fv3sar_grid_spec.nc
+
+#
```
Missing line of dashes here to stay consistent with style.

hu5970 replied:
Done

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,270 @@
+# Create a subdirectory in the current cycle's run di-
+# rectory for radar tten process.
+#
+#-----------------------------------------------------------------------
```
This comment is not true, and the next one has not been addressed, either.

hu5970 replied:
Change to Getting into working directory

christinaholtNOAA commented on scripts/exregional_refl2tten.sh:
```diff
@@ -0,0 +1,270 @@
+#-----------------------------------------------------------------------
+
+cp_vrfy ${fixdir}/fv3_akbk                               fv3_akbk
+cp_vrfy ${fixdir}/fv3_grid_spec                          fv3_grid_spec
```
These aren't background files, as the comment suggests.

christinaholtNOAA commented on scripts/exregional_run_fcst.sh:
```diff
@@ -329,13 +329,24 @@ of the current run directory (run_dir), where
+fi
+print_info_msg "$VERBOSE" "
+The forecast has BKTYPE $BKTYPE (1:cold start ; 2 cycling)"
+
```
What is BKTYPE=0? It's not reflected in this comment.

hu5970 replied:
This information was wrong. changed to ((1:cold start ; 0 cycling))

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+    print_info_msg "$VERBOSE"  "Bad GDAS member number ${nn}."
+  else
+    ls ${ENKF_FCST}/${enkfcstname}.mem${nn}.nemsio >> filelist03
+  fi
```
The check on file size isn't robust to a bad member001.

hu5970 replied:
Copied this from RAP/HRRR and we can improve it later when we have better idea.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+
+FV3SARPATH=${CYCLE_DIR}
+cp_vrfy ${fixdir}/fv3_akbk                               fv3_akbk
+cp_vrfy ${fixdir}/fv3_grid_spec                          fv3_grid_spec
```
These aren't really background files.

hu5970 replied:
I usually think everything on grid is for background but I added "grib configuration" for more accuracy description.

christinaholtNOAA replied:
Should this be grib or grid?

hu5970 replied:
It is grid. It is a typo in conversation. The code is right.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+  ulimit -a
+  APRUN="srun"
+  LD_LIBRARY_PATH="${UFS_WTHR_MDL_DIR}/FV3/ccpp/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
+  NCKS=/apps/nco/4.1.0-intel/bin/ncks
```
Does NCKS not exist in a module file? Can this be moved to the module file that we are managing here so we keep all the requirements in the same file?

hu5970 replied:
Only certain versions of the NCKS works on JET so far. I prefer to leave it as is now as I don't know if I can make right change now. Still need to figure out how to control the environmental for GSI runs.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+  fv3sar_bg_type=0
+fi
+
+# analysis time
```
This comment doesn't help with understanding what's happening.

hu5970 replied:
Change to "update times in coupler.res to current cycle time"

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+#-----------------------------------------------------------------------
+#
+# link observation files
+# copy observation files to working directory 
```
Please clean up this comment.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+#cp_vrfy ${USHDIR}/addtten.py                        addtten.py 
+#/scratch1/BMC/wrfruc/Samuel.Trahan/soft/anaconda2-5.3.1/bin/python3.7 addtten.py fv3_tracer
+#python addtten.py fv3_tracer
+#
```
Please remove the above commented code.

hu5970 replied:
I need to leave this part as a note for now. 

In the future, we need a better way to add then array.

christinaholtNOAA commented on scripts/exregional_run_analysis.sh:
```diff
@@ -0,0 +1,594 @@
+# Copy CRTM coefficient files based on entries in satinfo file
+for file in `awk '{if($1!~"!"){print $1}}' ./satinfo | sort | uniq` ;do
+#   ln_vrfy -sf -t ${CRTMFIX}/${file}.SpcCoeff.bin  .
+#   ln_vrfy -sf -t ${CRTMFIX}/${file}.TauCoeff.bin  .
```
Please remove commented code here and below.

hu5970 replied:
I removed some comments but need to leave the part for turning on bias correction in the future.

christinaholtNOAA commented on ush/config_defaults.sh:
```diff
@@ -352,12 +352,20 @@ WFLOW_LAUNCH_LOG_FN="log.launch_FV3LAM_wflow"
 # The length of each forecast, in integer hours.
 #
+# DA_CYCLE_INTERV:
+# Data assimilation cycle interval
```
This is mean to be an integer in hours? Maybe add that info?

hu5970 replied:
Added "in integer hours for now" because we only do hourly now. May be sub hourly will come soon.

christinaholtNOAA commented on ush/config_defaults.sh:
```diff
@@ -352,12 +352,20 @@ WFLOW_LAUNCH_LOG_FN="log.launch_FV3LAM_wflow"
+#
+# RESTART_INTERVAL:
+# Set up how often or which forecast hours the restart files should be
+# generated. This need to match with DA_CYCLE_INTERV.
```
There's a typo here that makes this unclear. Maybe "Frequency to write restart files from forecast in hours." You should also include information on the type of the value. It looks like a list, but it's unclear on what should be included in the list.

hu5970 replied:
Changed to
"+# Set up frequenency or list of the forecast hours that FV3 should
+# generate the restart files. The numbers need to match with DA_CYCLE_INTERV"

Hope this is more clear. It is very trick to setup the restart files for cycling.

christinaholtNOAA reviewed (commented):
I left another set of comments that were a bit more obvious once you addressed the last round. The code looks nice and there's a lot of hard work in here. These things will ultimately make it easier to maintain, and to get back into the authoritative repo later, when it's time to do that.

christinaholtNOAA commented:
@hu5970 I'm curious if you'd be willing to change all occurrences of the abbreviation ANAL to ANL for consistency sake across config/jobs/scripts. In general it's a good idea to try for more consistent notation for readability.

hu5970 commented:
I use "ANAL" to match "FCST".

I see only other scripts use "ANL" is "ush/get_extrn_mdl_file_dir_info.sh".
For now, I prefer "ANAL" for better alignment with other components.

hu5970 committed:
Improve code based on Chiristina's comments.

hu5970 commented:
Christina,

Thank you very much for your efforts to review and improve the code. All comments are good and significantly improved the code.

I just pushed the changes based on your comments this morning.

Please check,

Thanks,
Ming

hu5970 commented:
@christinaholtNOAA Any chance to more this forward before you take leave. I will have more PR coming for DA. Please try this PR is the start instead of final version of DA.

Thanks,
Ming

christinaholtNOAA reviewed (approved):
Ming, this looks good. Just another quick comment here, if you could address it.

hu5970 commented:
@christinaholtNOAA Thanks for reviewing during vacation. Have a good trip.

hu5970 commented:
@TrevorAlcott-NOAA  
Trevor,
 Please review or approve the PR. 
Thanks, Ming

christinaholtNOAA commented:
@hu5970 are you ready for me to merge now?

hu5970 commented:
@christinaholtNOAA Please merge the PR. Thanks, Ming
